File: Classes/THFinderFile/THFinderTools.m
// THFinderTools.m

#import "THFinderTools.h"
#import "THLog.h"

//--------------------------------------------------------------------------------------------------------------------------------------------
@implementation THFinderTools

+ (BOOL)displayFinderGetInfoOfPath:(NSString*)path
{
	if (path==nil || [[NSFileManager defaultManager] fileExistsAtPath:path]==NO)
		return NO;

	NSString *source=[NSString stringWithFormat:
					  @"tell application \"Finder\"\n"
					  	"set aPath to POSIX file \"%@\"\n"
					  	"open information window of item aPath\n"
					  	"activate\n"
					  "end tell\n",
					  path];

	NSDictionary *errorInfo=nil;
	NSAppleEventDescriptor *aed=[[[NSAppleScript alloc] initWithSource:source] executeAndReturnError:&errorInfo];
	if (aed==nil)
		THLogError(@"aed==nil errorInfo:%@",errorInfo);
	return aed!=nil?YES:NO;
}

+ (BOOL)isSystemFileNameInFolder:(NSString*)fileName mode:(NSInteger)mode
{
   const char *c_filename=fileName.fileSystemRepresentation;
   if (fileName==nil || c_filename==NULL)
	   return NO;
	return THIsFilenameHiddenExcluded(c_filename,strlen(c_filename),(int)mode)==true?YES:NO;
}

+ (BOOL)isVolumeHiddenFilename:(NSString*)filename
{
   const char *c_filename=filename.fileSystemRepresentation;
   if (filename==nil || c_filename==NULL)
	   return NO;
   return THIsVolumeHiddenFilename(c_filename,strlen(c_filename))==true?YES:NO;
}

+ (BOOL)isVolumeSystemHiddenFilename:(NSString*)filename mode:(NSInteger)mode
{
   const char *c_filename=filename.fileSystemRepresentation;
   if (filename==nil || c_filename==NULL)
	   return NO;
   return THIsVolumeSystemHiddenFilename(c_filename,strlen(c_filename),(int)mode)==true?YES:NO;
}

+ (NSArray*)finderLabels
{
	NSArray *labels=[[NSWorkspace sharedWorkspace] fileLabels];
	NSArray *colors=[[NSWorkspace sharedWorkspace] fileLabelColors];

	NSMutableArray *results=[NSMutableArray array];

	for (NSString *label in labels)
	{
		if (label==labels[0]) // le premier est "None"
			continue;

		NSImage *badge=[[NSImage alloc] initWithSize:NSMakeSize(16.0,16.0)];
		[badge lockFocus];
			NSBezierPath *bezierPath=[NSBezierPath bezierPathWithOvalInRect:NSMakeRect(2.0,2.0,12.0,12.0)];
			[(NSColor*)colors[[labels indexOfObject:label]] set];
			[bezierPath fill];
			[[NSColor colorWithCalibratedWhite:0.5 alpha:0.33] set];
			[bezierPath stroke];
		[badge unlockFocus];

		[results addObject:@{@"label":label,@"badge":badge}];
	}

	return [NSArray arrayWithArray:results];
}

+ (NSImage*)iconWithAliasMaskForType:(NSString*)type
{
	static NSImage *aliasMask=nil;
	if (aliasMask==nil)
	{
		aliasMask=[[[NSWorkspace sharedWorkspace] iconForFileType:NSFileTypeForHFSTypeCode(kAliasBadgeIcon)] copy];
		aliasMask.size=NSMakeSize(16.0,16.0);
	}

	NSImage *icon=type==nil?nil:[[[NSWorkspace sharedWorkspace] iconForFileType:type] copy];
	icon.size=NSMakeSize(16.0,16.0);

	NSImage *result=[[NSImage alloc] initWithSize:NSMakeSize(16.0,16.0)];
	[result lockFocus];
		[icon drawAtPoint:NSMakePoint(0.0,0.0) fromRect:NSMakeRect(0.0,0.0,16.0,16.0) operation:NSCompositingOperationSourceOver fraction:1.0];
		[aliasMask drawAtPoint:NSMakePoint(0.0,0.0) fromRect:NSMakeRect(0.0,0.0,16.0,16.0) operation:NSCompositingOperationSourceOver fraction:1.0];
	[result unlockFocus];

	return result;
}

@end
//--------------------------------------------------------------------------------------------------------------------------------------------

File: Classes/THFinderFile/THFinderTools.swift
// THFinderTools.swift

import Cocoa

//--------------------------------------------------------------------------------------------------------------------------------------------
@objc class THFinderTools: NSObject {

	@objc class func displayFinderGetInfo(ofPath path: String) -> Bool {
		if path == nil || FileManager.default.fileExists(atPath: path) == false {
			return false
		}

		var source = ""
		source += "tell application \"Finder\"\n"
		source += "set aPath to POSIX file \"\(path)\"\n"
		source += "open information window of item aPath\n"
		source += "activate\n"
		source += "end tell\n"

		var errorInfo: NSDictionary?
		if NSAppleScript(source: source)?.executeAndReturnError(&errorInfo) == nil {
			THLogError("executeAndReturnError == nil errorInfo:\(errorInfo)")
			return false
		}

		return true
	}

	@objc class func isSystemFilename(_ name: String, mode: Int) -> Bool {
		let c_name = (name as NSString).fileSystemRepresentation
		return THIsFilenameHiddenExcluded(c_name, strlen(c_name), Int32(mode))
	}

	@objc class func isVolumeHiddenFilename(_ name: String) -> Bool {
		let c_name = (name as NSString).fileSystemRepresentation
		return THIsVolumeHiddenFilename(c_name, strlen(c_name))
	}

	@objc class func isVolumeSystemHiddenFilename(_ name: String, mode: Int) -> Bool {
		let c_name = (name as NSString).fileSystemRepresentation
		return THIsVolumeSystemHiddenFilename(c_name, strlen(c_name), Int32(mode))
	}

	@objc class func finderLabels() -> [[String: Any]]? {
		let labels = NSWorkspace.shared.fileLabels
		let colors = NSWorkspace.shared.fileLabelColors

		var results = [[String: Any]]()

		for (idx, label) in labels.dropFirst().enumerated() { // le premier est "None"
			let badge = NSImage(size: NSSize(16.0, 16.0))
			
			badge.lockFocus()
				let bz = NSBezierPath(ovalIn: NSRect(2.0, 2.0, 12.0, 12.0))

				colors[idx].set()
				bz.fill()

				NSColor(calibratedWhite: 0.5, alpha: 0.33).set()
				bz.stroke()
			badge.unlockFocus()

			results.append(["label": label, "badge": badge])
		}

		return results
	}

//	@objc class func iconWithAliasMask(forType type: String) -> NSImage? {
//	private static let aliasMask = NSWorkspace.shared.icon(forFileType: NSFileTypeForHFSTypeCode(OSType(kAliasBadgeIcon))).th_copyAndResize(NSSize(16.0, 16.0))
//
//		let icon = NSWorkspace.shared.icon(forFileType: type).copy() as! NSImage
//		icon.size = NSMakeSize(16.0, 16.0)
//
//		NSImage *result=[[NSImage alloc] initWithSize:NSMakeSize(16.0,16.0)];
//		[result lockFocus];
//			[icon drawAtPoint:NSMakePoint(0.0,0.0) fromRect:NSMakeRect(0.0,0.0,16.0,16.0) operation:NSCompositingOperationSourceOver fraction:1.0];
//			[aliasMask drawAtPoint:NSMakePoint(0.0,0.0) fromRect:NSMakeRect(0.0,0.0,16.0,16.0) operation:NSCompositingOperationSourceOver fraction:1.0];
//		[result unlockFocus];
//
//		return result
//	}

}
//--------------------------------------------------------------------------------------------------------------------------------------------
